routes/logger.go

```diff
@@ -0,0 +1,69 @@
+package routes
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net/http"
+	"net/http/httptest"
+	"time"
+
+	"github.com/gocardless/draupnir/routes/chain"
+	"github.com/prometheus/common/log"
+)
+
+type key int
+
+const loggerKey key = 1
+
+func NewRequestLogger(logger log.Logger) chain.Middleware {
+	return func(next chain.Handler) chain.Handler {
+		return func(w http.ResponseWriter, r *http.Request) error {
+			// To capture the response, we replace the response writer with a response
+			// recorder.
+			recorder := httptest.NewRecorder()
+
+			scopedLogger := logger.With("http_request", r)
+
+			// Inject the logger into the request's context
+			ctx := r.Context()
```

Comment:
I'd probably prefer this as

```
ctx := context.WithValue(r.Context(), loggerKey, &scopedLogger)
r = r.WithContext(ctx)
```

Otherwise I tend to look for where we'll be later using the `ctx` local. Could image someone adding code further down and using `ctx` when the decorated context should be used for all things onward.

Reply:
Have gone with an approach that removes the local entirely:
```go
r = r.WithContext(context.WithValue(r.Context(), loggerKey, &scopedLogger))
```